lib/translator/read.cc

```diff
@@ -0,0 +1,196 @@
+#include "read.h"
+
+#ifdef __KERNEL__
+#include <linux/byteorder/generic.h>
+#else
+#include <netinet/in.h>
+#endif
+
+#include "absl/types/span.h"
+
+#include "common.h"
+
+namespace translator {
+
+namespace {  // anonymous namespace for helper functions
+
+// Section 5.3
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode BuildPrinfo(uint8_t rdprotect, uint8_t& prinfo) {
+  bool pract;     // Protection Information Action 1 bit
+  uint8_t prchk;  // Protection Information Check 3 bits
+
+  switch (rdprotect) {  // 3 bits
+    case 0b0:
+      pract = 0b1;
+      prchk = 0b111;
+      break;
+    case 0b001:
+    case 0b101:
+      pract = 0b0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0b0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0b0;
+      prchk = 0b000;
+      break;
+    case 0b100:
+      pract = 0b0;
+      prchk = 0b100;
+      break;
+    default:
+      // Should result in SCSI command termination with status: CHECK CONDITION,
+      //  sense key: ILLEGAL REQUEST, additional sense code: LLEGAL FIELD IN CDB
+      DebugLog("RDPROTECT with value %d has no translation to PRINFO",
+               rdprotect);
+      return StatusCode::kInvalidInput;
+  }
+
+  prinfo = prchk | (pract << 3);
+
+  return StatusCode::kSuccess;
+}
+
+// Builds NVMe cdw 12 for Read10, Read12, Read16 translations
+uint32_t BuildCdw12(uint32_t transfer_length, uint8_t prinfo, bool fua) {
+  // cdw12 nlb bits 15:00 (zero based field), printfo bits 29:26, fua bit 30
+  // Since NVMe has a protocol limit of 16 bits on transfer size,
+  // we can enforce a transfer limit of 16 bits on our SCSI device
```

Comment:
We should still have some sort of error checking to ensure that our length expectation is met.